### ISSUE TITLE
[Question] How to process refunds?

### ISSUE THREAD

**Issue by RosiersRobin** [mollie/laravel-cashier-mollie#109]
Hi,

I se how I can make refunds, but how can I make them process? It's in the webhook I guess, but there is no example on how to do this. Would it be to fetch te order, check if there are refunds and if so, just call the `handleProcessed()` on the Refund model?

**Comment by RosiersRobin** [comment]
Sorry, I know that mollie takes oooon test environments around 2 hours to get processed by their system. However, I aimed to know how I can see the changes in my application. Since I don't see those appearing...

**Comment by sandervanhooft** [comment]
Can you share a code sample so we can reproduce the issue?

**Comment by RosiersRobin** [comment]
So I have made a refund with;

```php
$order = Order::query()->find(x);
$order->refundCompletely();
```

That makes the refund in mollie. So that's good already.

This also makes table values in the refunds table;

![image](https://user-images.githubusercontent.com/5263188/182338147-86d5909c-6a70-46cf-8ea5-d99002333a40.png)

And also in the refund_items table, so that's okay.

Now, the status will never change, since there is never an update when it's been processed.

My webhook liiks like the following;

```php
final class WebhookController extends BaseWebhookController
{
    /**
     * @param Request $request
     * @return Response
     * @throws ApiException Only in debug mode
     */
    public function handleWebhook(Request $request): Response
    {
        $payment = $this->getPaymentById($request->get('id'));

        if ($payment) {
            $order = $this->getOrder($payment);

            if ($order && $order->mollie_payment_status !== $payment->status) {
                switch ($payment->status) {
                    case PaymentStatus::STATUS_PAID:
                        $order->handlePaymentPaid($payment);
                        break;
                    case PaymentStatus::STATUS_FAILED:
                        $order->handlePaymentFailed($payment);
                        break;
                    default:
                        break;
                }
            }
        }

        return new Response(null, 200);
    }

    /**
     * @param Payment $payment
     * @return Order|null
     */
    protected function getOrder(Payment $payment): ?Order
    {
        $order = Order::findByPaymentId($payment->id);

        if (!$order && isset($payment->metadata, $payment->metadata->temporary_mollie_payment_id)) {
            $order = Order::findByPaymentId($payment->metadata->temporary_mollie_payment_id);

            if ($order) {
                // Store the definite payment id.
                $order->update(['mollie_payment_id' => $payment->id]);
            }
        }

        return $order;
    }
}
```

Am I missing something here for refunds? I bet I am...

**Comment by sandervanhooft** [comment]
Hi @RosiersRobin ,

I am not sure why you have implemented your own controller for this as everything is already handled by Cashier for you...

But essentially the refunds are handled by the AftercareWebhookController.

Also note that - if I remember correctly - Mollie refunds receive no status updates in test mode. So there's no webhook request from Mollie after refund creation

**Comment by RosiersRobin** [comment]
Hi @sandervanhooft 

The reason why I made my own callbacks and such, is because I don't have users to do payments. For example, you can create a registration in the system, with catering attached to that, but that isn't connected to a user. the order is then connected to the registration... Which is not really possible, or am I wrong?

**Comment by RosiersRobin** [comment]
@sandervanhooft hi,

I have tried to change all the self-written code to the implementation of the package, however, when I test all of this, I always get a 404 on the first-pament webhook url (`/webhooks/mollie/first-payment`).

An order gets never made, even after triggering all the webhooks myself (since it's local development).

`cashier:run` also doesn't make any kind of orders. I followed the full docs, but it just doesn't seem to work.
